druid-shell/src/platform/mac/dialog.rs

```diff
@@ -142,3 +164,116 @@ pub(crate) unsafe fn build_panel(ty: FileDialogType, mut options: FileDialogOpti
     }
     panel
 }
+
+// AppKit has a build-in file format accessory view. However, this is only
+// displayed for `NSDocument` based apps. We have to construct our own `NSView`
+// hierachy to implement something similar.
+unsafe fn allowed_types_accessory_view(allowed_types: &[crate::FileSpec]) -> id {
+    // Build the View Structure required to have file format popup.
+    // This requires a container view, a label, a popup button,
+    // and some layout code to make sure it behaves correctly for
+    // fixed size and resizable views
+    let total_frame = NSRect::new(
+        NSPoint { x: 0.0, y: 0.0 },
+        NSSize {
+            width: 320.0,
+            height: 30.0,
+        },
+    );
+
+    let padding = 10.0;
+
+    // Prepare the label
+    let (label, label_size) = file_format_label();
+
+    // Place the label centered (similar to the popup button)
+    label.setFrameOrigin(NSPoint {
+        x: padding,
+        y: label_size.height / 2.0,
+    });
+
+    // Prepare the popup button
+    let popup_frame = NSRect::new(
+        NSPoint {
+            x: padding + label_size.width + padding,
+            y: 0.0,
+        },
+        NSSize {
+            width: total_frame.size.width - (padding * 3.0) - label_size.width,
+            height: total_frame.size.height,
+        },
+    );
+
+    let popup_button = file_format_popup_button(allowed_types, popup_frame);
+
+    // Prepare the container
+    let container_view: id = msg_send![class!(NSView), alloc];
+    let container_view: id = container_view.initWithFrame_(total_frame);
+
+    container_view.addSubview_(label);
+    container_view.addSubview_(popup_button);
+
+    container_view.autorelease()
+}
+
+const FileFormatPopoverTag: NSInteger = 10;
+
+unsafe fn file_format_popup_button(allowed_types: &[crate::FileSpec], popup_frame: NSRect) -> id {
+    let popup_button: id = msg_send![class!(NSPopUpButton), alloc];
+    let _: () = msg_send![popup_button, initWithFrame:popup_frame pullsDown:false];
+    for allowed_type in allowed_types {
+        let title = NSString::alloc(nil)
+            .init_str(allowed_type.name)
+            .autorelease();
+        msg_send![popup_button, addItemWithTitle: title]
+    }
+    let _: () = msg_send![popup_button, setTag: FileFormatPopoverTag];
+    popup_button.autorelease()
+}
+
+unsafe fn file_format_label() -> (id, NSSize) {
+    let label: id = msg_send![class!(NSTextField), new];
+    let _: () = msg_send![label, setBezeled:false];
+    let _: () = msg_send![label, setDrawsBackground:false];
+    // FIXME: As we have to roll our own view hierachy, we're not getting a translated
+    // title here. So we ought to find a way to translate this.
+    let title = NSString::alloc(nil).init_str("File Format:").autorelease();
```

Comment:
This is interesting. our architecture does not expect druid-shell to ever need to be locale aware; the idea is that localization will happen higher up in the stack, with correctly localized strings passed in to druid-shell where needed. I'm not sure how best to do that here, though. For the time being I'm happy to punt on this, but we will want to revisit it eventually..